ddtrace/internal/datadog/profiling/stack_v2/src/stack.cpp

```diff
@@ -0,0 +1,300 @@
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+#
+#if PY_VERSION_HEX >= 0x030c0000
+// https://github.com/python/cpython/issues/108216#issuecomment-1696565797
+#undef _PyGC_FINALIZED
+#endif
+
+#include <atomic>
+#include <chrono>
+#include <fcntl.h>
+#include <fstream>
+#include <iostream>
+#include <thread>
+#include <vector>
+
+#include "echion/config.h"
+#include "echion/interp.h"
+#include "echion/signals.h"
+#include "echion/stacks.h"
+#include "echion/state.h"
+#include "echion/threads.h"
+#include "echion/timing.h"
+
+#include "interface.hpp"
+
+// From https://gist.github.com/ichramm/3ffeaf7ba4f24853e9ecaf176da84566
+// TODO find a better thing and credit it
+bool
+utf8_check_is_valid(const char* str, int len)
+{
+    int n;
+    for (int i = 0; i < len; ++i) {
+        unsigned char c = (unsigned char)str[i];
+        // if (c==0x09 || c==0x0a || c==0x0d || (0x20 <= c && c <= 0x7e) ) n = 0; // is_printable_ascii
+        if (0x00 <= c && c <= 0x7f) {
+            n = 0; // 0bbbbbbb
+        } else if ((c & 0xE0) == 0xC0) {
+            n = 1; // 110bbbbb
+        } else if (c == 0xed && i < (len - 1) && ((unsigned char)str[i + 1] & 0xa0) == 0xa0) {
+            return false; // U+d800 to U+dfff
+        } else if ((c & 0xF0) == 0xE0) {
+            n = 2; // 1110bbbb
+        } else if ((c & 0xF8) == 0xF0) {
+            n = 3; // 11110bbb
+            //} else if (($c & 0xFC) == 0xF8) { n=4; // 111110bb //byte 5, unnecessary in 4 byte UTF-8
+            //} else if (($c & 0xFE) == 0xFC) { n=5; // 1111110b //byte 6, unnecessary in 4 byte UTF-8
+        } else {
+            return false;
+        }
+
+        for (int j = 0; j < n && i < len; ++j) { // n bytes matching 10bbbbbb follow ?
+            if ((++i == len) || (((unsigned char)str[i] & 0xC0) != 0x80)) {
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+class StackRenderer : public RendererInterface
+{
+    static PyObject* stack_sample_event_type;
+    static PyObject* ddframe_type;
+    void render_message(std::string_view msg) override { (void)msg; }
+
+    void render_thread_begin(PyThreadState* tstate,
+                             std::string_view name,
+                             microsecond_t wall_time,
+                             uintptr_t thread_id,
+                             unsigned long native_id) override
+    {
+        ddup_start_sample();
```

Comment:
If we want to allow the standardization of ddup to more code paths (allocations, exceptions...). We should consider removing the usage of globals.
This goes back to our previous discussion on ddup and how to expose an object.